File: backend/app/converter/structure.py
"""Structure detection: headings, paragraphs, tables, references via font-size heuristics."""

from __future__ import annotations

import re
import statistics

from app.models import PageBlock, Section, SemanticData


def detect_structure(blocks: list[PageBlock]) -> SemanticData:
    """Detect document structure from extracted blocks using font-size heuristics."""
    if not blocks:
        return SemanticData()

    font_sizes = [b.font_size for b in blocks if b.font_size > 0]
    if not font_sizes:
        return _fallback_structure(blocks)

    median_size = statistics.median(font_sizes)
    heading_threshold = median_size * 1.15

    sections: list[Section] = []
    references: list[str] = []
    current_section: Section | None = None
    in_references = False

    for block in blocks:
        text = block.text.strip()
        if not text:
            continue

        is_heading = (
            (block.font_size > heading_threshold or block.is_bold)
            and len(text) < 200
            and not text.endswith(".")
        )

        ref_header = re.match(r"^(references|bibliography|works cited)", text, re.IGNORECASE)
        if ref_header:
            in_references = True
            if current_section:
                sections.append(current_section)
            current_section = Section(
                title="References", level=1, content="", page=block.page
            )
            continue

        if in_references:
            references.append(text)
            if current_section:
                current_section.content += text + "\n"
            continue

        if is_heading:
            if current_section:
                sections.append(current_section)

            level = _estimate_level(block.font_size, font_sizes, heading_threshold)
            current_section = Section(
                title=text, level=level, content="", page=block.page
            )
        else:
            if current_section is None:
                current_section = Section(
                    title="Introduction", level=1, content="", page=block.page
                )
            current_section.content += text + " "

    if current_section:
        sections.append(current_section)

    for s in sections:
        s.content = s.content.strip()

    tables = _detect_tables(blocks)

    return SemanticData(sections=sections, tables=tables, references=references)


def _estimate_level(
    font_size: float, all_sizes: list[float], threshold: float
) -> int:
    large_sizes = sorted(set(s for s in all_sizes if s >= threshold), reverse=True)
    if not large_sizes:
        return 1
    for i, s in enumerate(large_sizes):
        if font_size >= s - 0.5:
            return i + 1
    return len(large_sizes)


def _detect_tables(blocks: list[PageBlock]) -> list[dict]:
    """Simple heuristic: sequences of lines with tab/multiple-space separation."""
    tables: list[dict] = []
    table_lines: list[str] = []
    table_page = 0

    for block in blocks:
        if "\t" in block.text or re.search(r" {3,}", block.text):
            if not table_lines:
                table_page = block.page
            table_lines.append(block.text)
        else:
            if len(table_lines) >= 2:
                tables.append({"page": table_page, "rows": table_lines})
            table_lines = []

    if len(table_lines) >= 2:
        tables.append({"page": table_page, "rows": table_lines})

    return tables


def _fallback_structure(blocks: list[PageBlock]) -> SemanticData:
    """When font metadata is unavailable, create a single section from all text."""
    full_text = " ".join(b.text for b in blocks)
    return SemanticData(
        sections=[
            Section(title="Document", level=1, content=full_text, page=1)
        ]
    )

File: desktop/frontend/src/core/services/adfConverter/structure.ts
/**
 * Structure detection using font-size heuristics.
 * Port of backend/app/converter/structure.py — detects headings, sections,
 * tables, and references from extracted text blocks.
 */
import type { PageBlock } from "@app/services/adfConverter/extractor";
import type { Section, SemanticData } from "@app/types/adf";

function median(values: number[]): number {
  if (values.length === 0) return 0;
  const sorted = [...values].sort((a, b) => a - b);
  const mid = Math.floor(sorted.length / 2);
  return sorted.length % 2 !== 0
    ? sorted[mid]
    : (sorted[mid - 1] + sorted[mid]) / 2;
}

function estimateLevel(
  fontSize: number,
  allSizes: number[],
  threshold: number
): number {
  const largeSizes = [
    ...new Set(allSizes.filter((s) => s >= threshold)),
  ].sort((a, b) => b - a);
  if (largeSizes.length === 0) return 1;
  for (let i = 0; i < largeSizes.length; i++) {
    if (fontSize >= largeSizes[i] - 0.5) return i + 1;
  }
  return largeSizes.length;
}

function detectTables(blocks: PageBlock[]): Array<{ page: number; rows: string[] }> {
  const tables: Array<{ page: number; rows: string[] }> = [];
  let tableLines: string[] = [];
  let tablePage = 0;

  for (const block of blocks) {
    if (block.text.includes("\t") || /\s{3,}/.test(block.text)) {
      if (tableLines.length === 0) tablePage = block.page;
      tableLines.push(block.text);
    } else {
      if (tableLines.length >= 2) {
        tables.push({ page: tablePage, rows: tableLines });
      }
      tableLines = [];
    }
  }
  if (tableLines.length >= 2) {
    tables.push({ page: tablePage, rows: tableLines });
  }
  return tables;
}

function fallbackStructure(blocks: PageBlock[]): SemanticData {
  const fullText = blocks.map((b) => b.text).join(" ");
  return {
    sections: [
      { title: "Document", level: 1, content: fullText, page: 1, summary: "" },
    ],
    tables: [],
    references: [],
  };
}

export function detectStructure(blocks: PageBlock[]): SemanticData {
  if (blocks.length === 0) {
    return { sections: [], tables: [], references: [] };
  }

  const fontSizes = blocks.filter((b) => b.font_size > 0).map((b) => b.font_size);
  if (fontSizes.length === 0) return fallbackStructure(blocks);

  const medianSize = median(fontSizes);
  const headingThreshold = medianSize * 1.15;

  const sections: Section[] = [];
  const references: string[] = [];
  let currentSection: Section | null = null;
  let inReferences = false;

  for (const block of blocks) {
    const text = block.text.trim();
    if (!text) continue;

    const isHeading =
      (block.font_size > headingThreshold || block.is_bold) &&
      text.length < 200 &&
      !text.endsWith(".");

    const refHeader = /^(references|bibliography|works cited)/i.test(text);
    if (refHeader) {
      inReferences = true;
      if (currentSection) sections.push(currentSection);
      currentSection = {
        title: "References",
        level: 1,
        content: "",
        page: block.page,
        summary: "",
      };
      continue;
    }

    if (inReferences) {
      references.push(text);
      if (currentSection) currentSection.content += text + "\n";
      continue;
    }

    if (isHeading) {
      if (currentSection) sections.push(currentSection);
      const level = estimateLevel(block.font_size, fontSizes, headingThreshold);
      currentSection = {
        title: text,
        level,
        content: "",
        page: block.page,
        summary: "",
      };
    } else {
      if (!currentSection) {
        currentSection = {
          title: "Introduction",
          level: 1,
          content: "",
          page: block.page,
          summary: "",
        };
      }
      currentSection.content += text + " ";
    }
  }

  if (currentSection) sections.push(currentSection);

  for (const s of sections) {
    s.content = s.content.trim();
  }

  const tables = detectTables(blocks);

  return { sections, tables, references };
}
